File: tests/cpp/TestMafComm.cpp
#include "maf/maf.hpp"

MAF_TEST_ACTION(test_bcast_string) {
    void run() {
        auto world = maf::MafComm::World;
        std::string str = std::string((world->rank + 3), 'm');

        if (world->rank == 0) {
            this->assert_equal(std::string("mmm"), str);
        }
        else {
            this->assert_not_equal(std::string("mmm"), str);
        }

        str = world->bcast(str);

        this->assert_equal(std::string("mmm"), str);
    }
};

MAF_TEST_ACTION(test_bcast_int) {
    void run() {
        auto world = maf::MafComm::World;
        int val = (world->rank + 3) * 3;

        if (world->rank == 0) {
            this->assert_equal(9, val);
        }
        else {
            this->assert_not_equal(9, val);
        }

        val = world->bcast(val);

        this->assert_equal(9, val);
    }
};

MAF_TEST_ACTION(test_bcast_int_vector) {
    void run() {
        auto world = maf::MafComm::World;
        std::vector<int> val(world->size - world->rank, (world->rank + 3) * 3);
        std::vector<int> expected {9, 9, 9};

        if (world->rank == 0) {
            this->assert_equal(expected, val);
        }
        else {
            this->assert_not_equal(expected, val);
        }

        val = world->bcast(val);
        this->assert_equal(expected, val);
    }
};

MAF_TEST_ACTION(test_bcast_float) {
    void run() {
        auto world = maf::MafComm::World;
        float val = (world->rank + 3.3) * 3;

        if (world->rank == 0) {
            this->assert_close(9.9f, val, 1e-5f);
        }
        else {
            this->assert_not_close(9.9f, val, 1e-2f);
        }

        val = world->bcast(val);
        this->assert_close(9.9f, val, 1e-5f);
    }
};

MAF_TEST_ACTION(test_bcast_float_vector) {
    void run() {
        auto world = maf::MafComm::World;
        std::vector<float> val(world->size - world->rank, (world->rank + 3) * 3.3);
        std::vector<float> expected {9.9f, 9.9f, 9.9f};

        if (world->rank == 0) {
            this->assert_close(expected, val, 1e-5f);
        }
        else {
            this->assert_not_equal(expected, val);
        }

        val = world->bcast(val);
        this->assert_close(expected, val, 1e-5f);
    }
};

MAF_TEST_ACTION(test_bcast_double) {
    void run() {
        auto world = maf::MafComm::World;
        double val = (world->rank + 3.3) * 3;

        if (world->rank == 0) {
            this->assert_close(9.9, val, 1e-5);
        }
        else {
            this->assert_not_close(9.9, val, 1e-2);
        }

        val = world->bcast(val);

        this->assert_close(9.9, val, 1e-5);
    }
};

MAF_TEST_ACTION(test_bcast_double_vector) {
    void run() {
        auto world = maf::MafComm::World;
        std::vector<double> val(world->size - world->rank, (world->rank + 3) * 3.3);
        std::vector<double> expected {9.9, 9.9, 9.9};

        if (world->rank == 0) {
            this->assert_close(expected, val, 1e-5);
        }
        else {
            this->assert_not_equal(expected, val);
        }

        val = world->bcast(val);
        this->assert_close(expected, val, 1e-5);
    }
};

MAF_TEST_MAIN()

File: tests/py/test_maf_comm.py
from mpi4py.MPI import COMM_WORLD 
import maf

@maf.test
def test_bcast_string(self):
    '''here is some doc for you'''
    maf.log('here.')
    world = maf.MafComm.World
    maf.log('here..')
    val = 'm' * (world.rank + 3)

    maf.log('here...')
    if world.rank == 0:
        self.assert_equal("mmm", val)
    else:
        self.assert_not_equal("mmm", val)
    maf.log('here....')

    val = world.bcast(val)
    maf.log('here.....')
    self.assert_equal("mmm", val)


@maf.test
def test_bcast_int(self):
    world = maf.MafComm.World
    val = (world.rank + 3) * 3

    if (world.rank == 0):
        self.assert_equal(9, val)
    else:
        self.assert_not_equal(9, val)

    val = world.bcast(val)
    self.assert_equal(9, val)


@maf.test
def test_bcast_float(self):
    world = maf.MafComm.World
    val = (world.rank + 3.3) * 3

    if (world.rank == 0):
        self.assert_close(9.9, val, 1e-5)
    else:
        self.assert_not_close(9.9, val, 1e-2)

    val = world.bcast(val)
    self.assert_close(9.9, val, 1e-5)


@maf.test
def test_bcast_double(self):
    world = maf.MafComm.World
    val = (world.rank + 3.3) * 3

    if (world.rank == 0):
        self.assert_close(9.9, val, 1e-5)
    else:
        self.assert_not_close(9.9, val, 1e-2)
    

    val = world.bcast(val)
    self.assert_close(9.9, val, 1e-5)


if __name__ == '__main__':
    maf.barrier('ERROR: these tests fail due to an error in SWiG. See https://github.com/swig/swig/issues/773')
    # controller = maf.TestController()
    # controller.start()
